File: app/src/at-hoangtran/java/com/asiantech/intern20summer1/intro/RainbowOfClarity.java
package com.asiantech.intern20summer1.intro;

public class RainbowOfClarity {
    public static void main(String[] args) {
        char symbol = 'g';
        String cell = "h1";
        String s = "aabbbc";
        int n = 35123;
        System.out.println(isDigit(symbol));
        System.out.println(lineEncoding(s));
        System.out.println(chessKnight(cell));
        System.out.println(deleteDigit(n));
    }

    static boolean isDigit(char symbol) {
//        Determine if the given character is a digit or not.
        // return(symbol>='0'&&symbol<='9');
        return (Character.isDigit(symbol));
    }

    static String lineEncoding(String s) {
//        Given a string,return its encoding defined as follows:
//
//        First, the string is divided into the least possible number of disjoint
//        substrings consisting of identical characters
//        for example, "aabbbc" is divided into["aa", "bbb", "c"]
//        Next, each substring with length greater than one is replaced with a concatenation of
//        its length and the repeating character
//        for example, substring "bbb" is replaced by "3b"
//        Finally, all the new strings are concatenated together in the same order and a new string
//        is returned.
        int so = 1;
        String string = "";
        for (int i = 0; i < s.length() - 1; i++) {
            if (s.charAt(i + 1) == s.charAt(i)) {
                so++;
            } else {
                if (so == 1) {
                    string += s.charAt(i);
                } else {
                    string += so + "" + s.charAt(i);
                    so = 1;
                }
            }
        }
        if (so == 1) {
            string += s.charAt(s.length() - 1);
        } else {
            System.out.print(string);
            string += so + "" + s.charAt(s.length() - 1);
        }
        return string;
    }

    static int chessKnight(String cell) {
//        Given a position of a knight on the standard chessboard, find the number of different
//        moves the knight can perform.
//
//                The knight can move to a square that is two squares horizontally and one square
//        vertically, or two squares vertically and one square horizontally away from it.The complete
//        move therefore looks like the letter L.Check out the image below to see all valid moves for
//        a knight piece that is placed on one of the central squares.
        int count = 0;
        if ((cell.charAt(0) - 1 >= 'a' && cell.charAt(0) - 1 <= 'h')
                && (cell.charAt(1) - 2 >= '1' && cell.charAt(1) - 2 <= '8')) {
            count++;
        }
        if ((cell.charAt(0) - 2 >= 'a' && cell.charAt(0) - 2 <= 'h')
                && (cell.charAt(1) - 1 >= '1' && cell.charAt(1) - 1 <= '8')) {
            count++;
        }
        if ((cell.charAt(0) + 1 >= 'a' && cell.charAt(0) + 1 <= 'h')
                && (cell.charAt(1) + 2 >= '1' && cell.charAt(1) + 2 <= '8')) {
            count++;
        }
        if ((cell.charAt(0) + 2 >= 'a' && cell.charAt(0) + 2 <= 'h')
                && (cell.charAt(1) + 1 >= '1' && cell.charAt(1) + 1 <= '8')) {
            count++;
        }
        if ((cell.charAt(0) - 1 >= 'a' && cell.charAt(0) - 1 <= 'h')
                && (cell.charAt(1) + 2 >= '1' && cell.charAt(1) + 2 <= '8')) {
            count++;
        }
        if ((cell.charAt(0) - 2 >= 'a' && cell.charAt(0) - 2 <= 'h')
                && (cell.charAt(1) + 1 >= '1' && cell.charAt(1) + 1 <= '8')) {
            count++;
        }
        if ((cell.charAt(0) + 1 >= 'a' && cell.charAt(0) + 1 <= 'h')
                && (cell.charAt(1) - 2 >= '1' && cell.charAt(1) - 2 <= '8')) {
            count++;
        }
        if ((cell.charAt(0) + 2 >= 'a' && cell.charAt(0) + 2 <= 'h')
                && (cell.charAt(1) - 1 >= '1' && cell.charAt(1) - 1 <= '8')) {
            count++;
        }
        return count++;
    }

    static int deleteDigit(int n) {
//        Given some integer, find the maximal number you can obtain by deleting exactly one digit of
//        the given number.
                String s = Integer.toString(n);
        int max = Integer.valueOf(s.substring(1));
        for (int i = 1; i < s.length(); i++) {
            if (i == s.length() - 1) {
                if (max < Integer.valueOf(s.substring(0, i))) {
                    max = Integer.valueOf(s.substring(0, i));
                }
            } else if (max < Integer.valueOf(s.substring(0, i) + s.substring(i + 1, s.length()))) {
                max = Integer.valueOf(s.substring(0, i) + s.substring(i + 1, s.length()));
            }

            System.out.println(max);
        }
        return max;
    }

}

File: app/src/at-longphan/java/com/asiantech/intern20summer1/introKotlin/RainbowOfClarity.kt
package com.asiantech.intern20summer1.introKotlin

fun main() {
    //RUN main() with Coverage
    var obj: RainbowOfClarity = RainbowOfClarity()
    println("Ex 48: " + obj.isDigit('0'))

    println("Ex 49: " + obj.lineEncoding("aabbbc"))

    println("Ex 50: " + obj.chessKnight("a1"))

    println("Ex 51: " + obj.deleteDigit(152))
}

class RainbowOfClarity {
    /**
     * Determine if the given character is a digit or not.
     */
    fun isDigit(symbol: Char): Boolean {
        val symbolRange = 48..57
        return symbol.toInt() in symbolRange
    }

    fun lineEncoding(s: String): String? {
        /**
         * Given a string, return its encoding defined as follows:

        First, the string is divided into the least possible number of disjoint substrings
        consisting of identical characters
        for example, "aabbbc" is divided into ["aa", "bbb", "c"]
        Next, each substring with length greater than one is replaced with a concatenation of
        its length and the repeating character
        for example, substring "bbb" is replaced by "3b"
        Finally, all the new strings are concatenated together in the same order and a new string is returned.
         */
        var size = 1
        for (i in 1 until s.length) {
            if (s[i] != s[i - 1]) {
                size++
            }
        }
        val letters = arrayOfNulls<Char>(size)
        letters[0] = s[0]
        var indexLetters = 0
        for (i in 1 until s.length) {
            if (s[i] == letters[indexLetters]) {
            } else {
                indexLetters++
                letters[indexLetters] = s[i]
            }
        }
        val repeats = IntArray(size)
        repeats[0] = 1
        var indexRepeats = 0
        for (i in 1 until s.length) {
            if (s[i] == s[i - 1]) {
                repeats[indexRepeats]++
            } else {
                indexRepeats++
                repeats[indexRepeats]++
            }
        }
        var result = ""
        for (i in letters.indices) {
            result += if (repeats[i] == 1) {
                (letters[i]!!).toString()
            } else {
                repeats[i].toString() + (letters[i]!!).toString()
            }
        }
        return result
    }

    fun chessKnight(cell: String): Int {
        /**
         * Given a position of a knight on the standard chessboard, find the number of
         * different moves the knight can perform.

        The knight can move to a square that is two squares horizontally and one square vertically,
        or two squares vertically and one square horizontally away from it.
        The complete move therefore looks like the letter L. Check out the image below
        to see all valid moves for a knight piece that is placed on one of the central squares.
         */
        var moves = 8
        if (cell[0] == 'b' || cell[0] == 'g') moves -= 2
        if (cell[1] == '2' || cell[1] == '7') moves -= 2
        if (cell[0] == 'a' || cell[0] == 'h') moves /= 2
        if (cell[1] == '1' || cell[1] == '8') moves /= 2
        return moves
    }

    fun deleteDigit(n: Int): Int {
        /**
         * Given some integer, find the maximal number you can obtain by deleting exactly one digit of the given number.
         */
        val stringN = n.toString()
        val length = stringN.length
        val x = IntArray(length)
        for (i in 0 until length) {
            val stringDeleted = stringN.substring(0, i) + stringN.substring(i + 1)
            x[i] = Integer.valueOf(stringDeleted)
        }
        var max = 0
        for (item in x) {
            max = if (item > max) item else max
        }
        return max
    }
}